File: webapp/server-node/src/logger.ts
/**
 * logger.ts — CROSS-CUTTING: application logging with mode + level control
 * ===========================================================================
 * WHAT THIS SOLVES
 * ----------------
 * The owner asked for three observable behaviours:
 *   1. Ability to SEE the app's logs while running.
 *   2. Ability to ENABLE/DISABLE logging overall.
 *   3. Separate CUSTOMER logging (what we ship/collect in production) from
 *      DEV logging (verbose, human-readable during development).
 *
 * HOW IT WORKS
 * ------------
 * Two environment variables control everything — no code changes needed:
 *
 *   LOG_LEVEL  = off | error | warn | info | debug     (default: info)
 *                'off' disables ALL output (requirement 2).
 *                'debug' adds developer-only detail (requirement 3, dev side).
 *
 *   LOG_FORMAT = dev | json                            (default: depends on NODE_ENV)
 *                dev  → human-readable coloured lines for terminals (dev mode)
 *                json → one-line machine-parseable JSON per event; this is the
 *                       "customer" format shipped to production log collectors
 *                       (CloudWatch/Datadog/ELK parse it natively).
 *
 * USAGE
 * -----
 *   import { makeLogger } from './logger.js';
 *   const log = makeLogger('issues');       // scope shown in every line
 *   log.info('issue created', { id });      // visible at info+
 *   log.debug('stage rule eval', {...});    // visible ONLY at debug
 *   log.warn(...); log.error(...);
 *
 * DESIGN RULES
 * ------------
 * - Logging NEVER throws and NEVER blocks business logic (all output is
 *   wrapped; a broken stdout must not fail an HTTP request).
 * - Audit trail ≠ logging: the append-only audit_log (src/audit.ts) records
 *   security-relevant actions to the DATABASE and runs regardless of log
 *   level. Logs are ephemeral diagnostics; audit is compliance evidence.
 * - Secrets (passwords, tokens) must never be passed as metadata.
 *
 * REQUIREMENT TRACEABILITY
 * ------------------------
 *   - Owner review "enable/disable logs, customer vs dev logging" (2026-08-25)
 *   - docs/ARCHITECTURE_EVOLUTION_PLAN.md §6 Security (no secrets in logs)
 */

/** Ordered severities; 'off' sits below everything so nothing prints. */
type Level = 'off' | 'error' | 'warn' | 'info' | 'debug';

const LEVEL_ORDER: Record<Level, number> = {
  off: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4,
};

/** Read once at boot; changing env vars requires restart (documented behaviour). */
const CURRENT_LEVEL: Level = (process.env.LOG_LEVEL as Level) || 'info';
const CURRENT_FORMAT: 'dev' | 'json' =
  (process.env.LOG_FORMAT as 'dev' | 'json') ||
  (process.env.NODE_ENV === 'production' ? 'json' : 'dev');

/** ANSI colour codes — terminal-only nicety, stripped naturally by collectors. */
const COLORS = {
  error: '\x1b[31m', // red
  warn: '\x1b[33m', // yellow
  info: '\x1b[36m', // cyan
  debug: '\x1b[90m', // grey
  reset: '\x1b[0m',
} as const;

/** ISO timestamp helper (UTC, collector-friendly). */
function timestamp(): string {
  return new Date().toISOString();
}

/** Serialise metadata safely: undefined → omitted; circular refs tolerated. */
function metaString(meta?: Record<string, unknown>): string {
  if (!meta || Object.keys(meta).length === 0) return '';
  try {
    return ' ' + JSON.stringify(meta);
  } catch {
    return ' [unserializable metadata]';
  }
}

/**
 * Create a scoped logger. One instance per module keeps output greppable:
 * `grep '"scope":"authz"'` finds every permission decision in production.
 */
export interface Logger {
  /** Severity-gated emitters; each accepts a message + optional metadata. */
  error(msg: string, meta?: Record<string, unknown>): void;
  warn(msg: string, meta?: Record<string, unknown>): void;
  info(msg: string, meta?: Record<string, unknown>): void;
  debug(msg: string, meta?: Record<string, unknown>): void;
}

export function makeLogger(scope: string): Logger {
  /** Core emitter shared by the four severity methods. */
  function emit(level: Exclude<Level, 'off'>, msg: string, meta?: Record<string, unknown>) {
    // Gate 1 — global kill switch / level filter ('off' disables EVERYTHING).
    if (LEVEL_ORDER[CURRENT_LEVEL] === 0) return;
    if (LEVEL_ORDER[level] > LEVEL_ORDER[CURRENT_LEVEL]) return;

    if (CURRENT_FORMAT === 'json') {
      // Customer/production format: stable keys for log queries & alerts.
      const line = JSON.stringify({ ts: timestamp(), level, scope, msg, ...meta });
      write(line);
    } else {
      // Developer format: readable, coloured, scoped.
      const color = COLORS[level];
      write(
        `${color}[${timestamp()}] ${level.toUpperCase().padEnd(5)} [${scope}]${COLORS.reset} ${msg}${metaString(meta)}`
      );
    }
  }

  /** Isolate stdout failures so logging can never crash the app. */
  function write(line: string) {
    try {
      console.log(line);
    } catch {
      /* ignore — logging must never throw */
    }
  }

  return {
    error: (m, meta) => emit('error', m, meta),
    warn: (m, meta) => emit('warn', m, meta),
    info: (m, meta) => emit('info', m, meta),
    debug: (m, meta) => emit('debug', m, meta),
  };
}

/** Boot-time banner proving the chosen configuration (visible when enabled). */
export function logBootConfig() {
  const log = makeLogger('boot');
  log.info('logging configured', { level: CURRENT_LEVEL, format: CURRENT_FORMAT });
}

File: webapp/server-rust/src/logger.rs
//! logger.rs — CROSS-CUTTING: application logging with mode + level control
//! ===========================================================================
//! EXACT behavioural mirror of `server-node/src/logger.ts` (docs/LOGGING_GUIDE.md).
//!
//! OWNER REQUIREMENT (why this exists)
//! -----------------------------------
//!  1. SEE the app's logs while running.
//!  2. ENABLE/DISABLE logging overall.
//!  3. Separate CUSTOMER logging (production, machine-parseable) from DEV
//!     logging (verbose, human-readable).
//!
//! CONTROL MODEL — two environment variables, read once at boot:
//!   LOG_LEVEL  = off | error | warn | info | debug   (default: info)
//!                'off' disables ALL output.
//!   LOG_FORMAT = dev | json                          (default: dev)
//!                dev  → coloured terminal lines
//!                json → one machine-parseable object per line for collectors
//!
//! DESIGN RULES
//! ------------
//! - Logging NEVER panics and never blocks business logic (writes are wrapped;
//!   a broken stdout must not fail an HTTP request).
//! - Audit ≠ logging: the append-only audit trail lives in the Db regardless
//!   of log level. Logs are ephemeral diagnostics; audit is compliance evidence.
//!
//! REQUIREMENT TRACEABILITY
//! ------------------------
//!   - Owner review "enable/disable logs, customer vs dev logging"
//!   - docs/LOGGING_GUIDE.md · docs/ARCHITECTURE_EVOLUTION_PLAN.md §6

use std::sync::atomic::{AtomicU8, Ordering};

/// Ordered severities; OFF sits below everything so nothing prints.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum Level {
    Off = 0,
    Error = 1,
    Warn = 2,
    Info = 3,
    Debug = 4,
}

impl Level {
    fn from_env() -> Level {
        match std::env::var("LOG_LEVEL").unwrap_or_default().to_lowercase().as_str() {
            "off" => Level::Off,
            "error" => Level::Error,
            "warn" => Level::Warn,
            "debug" => Level::Debug,
            _ => Level::Info,
        }
    }
    fn as_str(&self) -> &'static str {
        match self { Level::Off => "off", Level::Error => "error", Level::Warn => "warn", Level::Info => "info", Level::Debug => "debug" }
    }
}

/// Global gate — atomic so any thread can check without locking.
static CURRENT_LEVEL: AtomicU8 = AtomicU8::new(3); // info
static JSON_FORMAT: AtomicU8 = AtomicU8::new(0); // 0 = dev, 1 = json

/// Read env config once at boot; also prints the boot banner (when enabled).
pub fn init_from_env() {
    let level = Level::from_env();
    CURRENT_LEVEL.store(level as u8, Ordering::Relaxed);
    let format = std::env::var("LOG_FORMAT").unwrap_or_default().to_lowercase();
    JSON_FORMAT.store(if format == "json" { 1 } else { 0 }, Ordering::Relaxed);
    make_logger("boot").info(
        "logging configured",
        &[("level", level.as_str()), ("format", if format == "json" { "json" } else { "dev" })],
    );
}

fn enabled(level: Level) -> bool {
    (level as u8) <= CURRENT_LEVEL.load(Ordering::Relaxed) && CURRENT_LEVEL.load(Ordering::Relaxed) != 0
}

/// ISO-8601 UTC timestamp without external crates (seconds precision is fine
/// for logs; collectors add their own ingestion timestamps anyway).
fn timestamp() -> String {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default();
    let secs = now.as_secs();
    // Civil-from-days algorithm (Howard Hinnant) — no chrono dependency needed.
    let days = secs / 86_400;
    let rem = secs % 86_400;
    let (h, m, s) = (rem / 3600, (rem % 3600) / 60, rem % 60);
    let z = days as i64 + 719_468;
    let era = z.div_euclid(146_097);
    let doe = z.rem_euclid(146_097);
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let mo = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if mo <= 2 { y + 1 } else { y };
    format!("{y:04}-{mo:02}-{d:02}T{h:02}:{m:02}:{s:02}Z")
}

/// A scoped logger — one instance per module keeps output greppable:
/// `grep '"scope":"authz"'` finds every permission decision in production.
pub struct Logger {
    scope: &'static str,
}

pub fn make_logger(scope: &'static str) -> Logger {
    Logger { scope }
}

impl Logger {
    pub fn error(&self, msg: &str, meta: &[(&str, &str)]) { self.emit(Level::Error, msg, meta); }
    pub fn warn(&self, msg: &str, meta: &[(&str, &str)]) { self.emit(Level::Warn, msg, meta); }
    pub fn info(&self, msg: &str, meta: &[(&str, &str)]) { self.emit(Level::Info, msg, meta); }
    pub fn debug(&self, msg: &str, meta: &[(&str, &str)]) { self.emit(Level::Debug, msg, meta); }

    fn emit(&self, level: Level, msg: &str, meta: &[(&str, &str)]) {
        if !enabled(level) {
            return; // Gate: global kill-switch / level filter.
        }
        let line = if JSON_FORMAT.load(Ordering::Relaxed) == 1 {
            // Customer/production format: stable keys for log queries & alerts.
            let mut obj = format!("{{\"ts\":\"{}\",\"level\":\"{}\",\"scope\":\"{}\",\"msg\":\"{}\"", timestamp(), level.as_str(), self.scope, msg);
            for (k, v) in meta {
                obj.push_str(&format!(",\"{k}\":\"{v}\""));
            }
            obj.push('}');
            obj
        } else {
            // Developer format: readable, scoped (colours omitted for portability
            // across Windows consoles that may not honour ANSI by default).
            let mut s = format!("[{}] {} [{}] {}", timestamp(), level.as_str().to_uppercase(), self.scope, msg);
            for (k, v) in meta {
                s.push_str(&format!(" {k}={v}"));
            }
            s
        };
        // Never panic on stdout failure — logging must never crash the app.
        println!("{line}");
    }
}
